go/plugins/dotprompt/dotprompt.go

```diff
@@ -321,3 +330,125 @@ func sortSchemaSlices(s *jsonschema.Schema) {
 		sortSchemaSlices(s.Items)
 	}
 }
+
+// WithTools adds tools to the prompt.
+func WithTools(tools ...ai.Tool) PromptOption {
+	return func(p *Prompt) error {
+		var toolSlice []ai.Tool
+		toolSlice = append(toolSlice, tools...)
+		p.Tools = toolSlice
+		return nil
+	}
+}
+
+// WithDefaultConfig adds default model configuration.
+func WithDefaultConfig(config *ai.GenerationCommonConfig) PromptOption {
+	return func(p *Prompt) error {
+		p.Config.GenerationConfig = config
+		return nil
+	}
+}
+
+// WithInputType uses the type provided to derive the input schema.
+func WithInputType(input any) PromptOption {
+	return func(p *Prompt) error {
+		if p.Config.InputSchema != nil {
+			return errors.New("dotprompt.WithInputType: cannot set both input schema and input type")
+		}
+
+		r := &jsonschema.Reflector{
+			AllowAdditionalProperties: false,
+			DoNotReference:            true,
+		}
+		p.Config.InputSchema = r.Reflect(input)
+		return nil
+	}
+}
+
+// WithInputSchema adds the schema for input to the prompt.
+func WithInputSchema(schema *jsonschema.Schema) PromptOption {
+	return func(p *Prompt) error {
+		if p.Config.InputSchema != nil {
+			return errors.New("dotprompt.WithInputSchema: cannot set both input schema and input type")
+		}
+
+		p.Config.InputSchema = schema
+		return nil
+	}
+}
+
+// WithOutputType uses the type provided to derive the output schema.
+func WithOutputType(output any) PromptOption {
+	return func(p *Prompt) error {
+		if p.Config.OutputSchema != nil {
+			return errors.New("dotprompt.WithOutputType: cannot set both output schema and output type")
+		}
+
+		r := &jsonschema.Reflector{
+			AllowAdditionalProperties: false,
+			DoNotReference:            true,
+		}
+		p.Config.OutputSchema = r.Reflect(output)
+		p.Config.OutputFormat = ai.OutputFormatJSON
+		return nil
+	}
+}
+
+// WithOutputSchema adds the schema for output to the prompt.
+func WithOutputSchema(schema *jsonschema.Schema) PromptOption {
+	return func(p *Prompt) error {
+		if p.Config.OutputSchema != nil {
+			return errors.New("dotprompt.WithOutputSchema: cannot set both output schema and output type")
+		}
+
+		p.Config.OutputSchema = schema
+		p.Config.OutputFormat = ai.OutputFormatJSON
+		return nil
+	}
+}
+
+// WithOutputFormat adds the desired output format to the prompt
+func WithOutputFormat(format ai.OutputFormat) PromptOption {
+	return func(p *Prompt) error {
+		p.Config.OutputFormat = format
+		return nil
+	}
+}
+
+// WithDefaults adds default input variable values
+func WithDefaults(variableDefaults map[string]any) PromptOption {
```

Comment:
If we combine WithInputType and WithDefaults, then if you choose to set inputSchema, you would have no way to pass in defaults at the prompt level? I'm assuming defaults work right now also when setting a schema directly over a type?

We could rename WithDefaults to WithDefaultVariables, then it would match the WithVariables on generate, and the WithDefaultModel/WithModel and WithDefaultModelName/WithModelName naming we went with yesterday.

Reply:
I'm not sure that `WithInputSchema` is all that important to have, for example, we don't offer that option for model generation -- it's input type only (despite the names there saying "with schema", which I think we should rename). JSON schema is kind of an implementation detail of Genkit, Go users will be using Go types.